rfcs/2020-06-12-2768-batch-and-buffer-rework.md

```diff
@@ -0,0 +1,129 @@
+   flow rate is high enough to avoid triggering a batch timeout.
+
+## Guide-Level Proposal
+
+For sinks that use the `BatchSink` framework (`PartitionBatchSink` is
+similar), events move along the following path:
+
+1. Each event is optionally encoded into a more suitable format for
+   buffering (using `with_flat_map` on the batch sink, ref `type
+   Batch::Input`).
+2. The encoded item is pushed into a buffer (in
+   `BatchSink::start_send`).
+3. When that buffer is full, as determined by being at or above the
+   configured batch size, it is collected into a batch (in
+   `BatchSink::poll_complete`).
+4. The batch is turned into a request (in `trait Service::call`) which
+   will do the final encoding and serialization.
+
+There exist the following buffer types:
+
+- `Buffer` offers (gzip) compression, but requires the events be
+  serialized before insertion and that serialization must include a
+  terminator. Events are stored and output as one large `Vec<u8>`, which
+  allows for easy computation of both sizes.
+- `JsonArrayBuffer` internally serializes events into "raw" JSON, which
+  may be then added to another structure after output. Events are stored
+  as `serde_json::value::RawValue` which is a newtype over a boxed
+  string, which allows for maintaining a running total of both sizes.
+- `MetricBuffer` stores metrics in their structured form, without
+  encoding. Since the target serialization is unknown, this buffer
+  cannot know the total byte size.
+- `PartitionBuffer<T, K>` is a meta-buffer, that partitions events into
+  multiple buffers of type `T`.
+- `Vec<T>` can only count events, not bytes, as it is agnostic to the
+  size of `T`.
+
+We want to get to the point where as many batching sinks as possible can
+be configured with either a byte limit or an event limit, or
+both. Additionally, the batch limits must be inviolable, so that no
+requests are ever sent that exceed size limits. To accomplish this the
+following is necessary:
+
+### Unified Batch Configuration
+
+The split configuration types for bytes and events will be rolled into a
+single configuration combining both attributes. The `max_size` element
+is retained for backwards compatibility, and each sink will use it as a
+default for the maximum bytes or events depending on the previous
+configuration mode.
+
+```rust
+struct BatchConfig {
+    max_bytes: Option<usize>,
+    max_events: Option<usize>,
+    max_size: Option<usize>,
+    timeout_secs: Option<u64>,
+}
+```
+
+### Failible Batch Insertion
+
+The current batch buffers always allow events to be inserted, and then
+check separately if the buffer is "big enough" to be sent. This will be
+changed into a new signature that indicates if the insertion would push
+the buffer over its size limits.
+
+```rust
+#[must_use]
+enum PushResult<E> {
+    Ok,
+    Full(E),
+}
+
+trait Batch {
+    fn push(&mut self, event: Self::Input) -> PushResult<Self::Input>;
```

Comment:
Seems like this could return an `Option<Input>` and avoid the need for a specialized type?

Reply:
I considered this, but didn't like the semantics. With a return of `Option<Self::Input>`, if the push worked, it would return `None`, and a failure would be indicated with `Some(_)`. That's unusual for an option.

Reply:
Yeah, I was thinking how `HashMap::insert` works I guess. :)

Reply:
To contrast, `HashMap::insert` always succeeds, the return value is just a matter of "do I care what was in there before I inserted". This one may fail, and so the return value must not be ignored, ruling out `Option`. I considered `Result<(), Self::Input>`, but the input isn't an error type, so that doesn't make sense. Thus, a custom type.